### PR TITLE
Fast QR Factorization on GPU

### PR DESCRIPTION
**Pull request by mj023**
There is an Error in the Jax QR Factorization that leads to slow execution for batches of small matrices when calculating the gradient. This PR tries to fix this, by implementing our own QR Factorization that runs much faster when using a GPU.

- [x] Implement faster QR
- [x] Identify the Jax backend and select the right QR Factorization

### CODE REVIEW

File: tests/test_qr.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import jax
+import jax.numpy as jnp
+import numpy as np
+from numpy.testing import assert_array_almost_equal as aaae
+from skillmodels.qr import qr_gpu
+
+def test_qr():
+    factorized = np.random.uniform(low=-1, high=3, size=(7, 7))
```
Comment (hmgaudecker) [inline tests/test_qr.py:8]: Please fix the seed (using modern numpy) -- just in case we ever get a failure, we want to be able to reproduce it.

File: tests/test_qr.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        q,r = jnp.linalg.qr(A)
+        return jnp.sum(r) + jnp.sum(q)
+    def f_gpu(A):
+        q,r = qr_gpu(A)
+        return jnp.sum(r) + jnp.sum(q)
+    grad_qr_jax = jax.grad(f_jax)
+    grad_qr_gpu = jax.grad(f_gpu)
+    grad_gpu = grad_qr_gpu(cov)
+    grad_jax = grad_qr_jax(cov)
+    aaae(q_gpu, q_jax)
```
Comment (hmgaudecker) [inline tests/test_qr.py:22]: Please make 4 tests out of this, setting up `factorized` and `cov` via a fixture. In case the first one fails, we want to see the pattern of failures.

**Review by hmgaudecker — COMMENTED** [review]
Looks great, thanks!

Haven't looked at the algorithm yet.

In addition to the couple of suggestions on the tests, can you please:
- Set [`endogenize_investments`](https://github.com/OpenSourceEconomics/skillmodels/tree/endogenize_investments) as the target branch.
- Run the pre-commit hooks on your changes (`pixi global install pre-commit`, `pre-commit run --files src/skillmodels/kalman_filters.py` etc.)

File: tests/test_qr.py (hunk 2 above)
Comment (mj023) [reply tests/test_qr.py]: I split this into 3 tests, I realized the last one was a duplicate.

**Comment by hmgaudecker** [comment]
Thanks a lot!
